components/server/src/workspace/workspace-starter.ts

```diff
@@ -530,54 +530,86 @@ export class WorkspaceStarter {
                 ...user,
             };
 
-            // choose a cluster and start the instance
-            let resp: StartWorkspaceResponse.AsObject | undefined = undefined;
-            let retries = 0;
-            try {
-                for (; retries < MAX_INSTANCE_START_RETRIES; retries++) {
-                    resp = await this.tryStartOnCluster({ span }, startRequest, euser, workspace, instance);
-                    if (resp) {
-                        break;
+            const ideUrlPromise = new Deferred<string>();
+
+            (async () => {
+                try {
+                    // choose a cluster and start the instance
+                    let resp: StartWorkspaceResponse.AsObject | undefined = undefined;
+                    let retries = 0;
+                    try {
+                        for (; retries < MAX_INSTANCE_START_RETRIES; retries++) {
+                            resp = await this.tryStartOnCluster({ span }, startRequest, euser, workspace, instance);
+                            if (resp) {
+                                break;
+                            }
+                            await new Promise((resolve) =>
+                                setTimeout(resolve, INSTANCE_START_RETRY_INTERVAL_SECONDS * 1000),
+                            );
+                        }
+                    } catch (err) {
+                        await this.failInstanceStart({ span }, err, workspace, instance);
+                        throw new StartInstanceError("startOnClusterFailed", err);
+                    }
+
+                    if (!resp) {
+                        const err = new Error("cannot start a workspace because no workspace clusters are available");
+                        await this.failInstanceStart({ span }, err, workspace, instance);
+                        throw new StartInstanceError("clusterSelectionFailed", err);
                     }
-                    await new Promise((resolve) => setTimeout(resolve, INSTANCE_START_RETRY_INTERVAL_SECONDS * 1000));
+                    increaseSuccessfulInstanceStartCounter(retries);
+
+                    span.log({ resp: resp });
```

Comment:
Do we want to log some extra text for humans? Just logging the response can make it harder to filter on these responses.

Reply:
Also, can we attach the context of the request? - instanceId, userID, workspaceID?

Reply:
I didn't touch that statement as part of this PR. It's the log on the trace so, I'm not even sure this is being used at all atm.

Reply:
Sorry, my bad. I read it as a log statement, not a trace statement.

Reply:
But I guess if we do want to compare which of the two returns more often, we should have a log here